File: leetcode/java/0014_longest_common_prefix.java
class Solution {
    public String longestCommonPrefix(String[] strs) {
        char currChar = 'a';

        if (strs[0].length() != 0) {
            currChar = strs[0].charAt(0);
        }

        int j = 0, i = 0;
        StringBuilder res = new StringBuilder();

        while (true) {
            if (j > strs[i].length() - 1) {
                break;
            }
            if (i == 0) {
                currChar = strs[i].charAt(j);
            }
            if (i > 0 && strs[i].charAt(j) != currChar) {
                break;
            }
            if (i == strs.length - 1 && strs[i].charAt(j) == currChar) {
                res.append(currChar);
            }

            if (i == strs.length - 1) {
                i = 0;
                j++;
            } else {
                i++;
            }
        }

        return res.toString();
    }
}

File: leetcode/cpp/0014_longest_common_prefix.cpp
class Solution {
public:
    string commonPrefix(string left, string right) {
        int minStr = std::min(left.size(), right.size());
        
        for (int i = 0; i < minStr; i++) {
            if (left[i] != right[i]) {
                return left.substr(0, i);
            }
        }
        
        return left.substr(0, minStr);
    }
    
    string longestCommonPrefixHelper(vector<string>& strs, int l, int r) {
        string lcpLeft, lcpRight;
        
        if (l == r) {
            return strs[l];
        }
        else if (l != r) {
            int mid = (l + r) / 2;
            lcpLeft = longestCommonPrefixHelper(strs, l, mid);
            lcpRight = longestCommonPrefixHelper(strs, mid + 1, r);
        }
        
        return commonPrefix(lcpLeft, lcpRight);
    }
    
    string longestCommonPrefix(vector<string>& strs) {
        if (strs.size() == 0) {
            return "";
        }
        
        return longestCommonPrefixHelper(strs, 0, strs.size() - 1);
    }
};